cirq/ops/common_channels.py

```diff
@@ -0,0 +1,188 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Quantum channels that are commonly used in the literature."""
+
+from typing import Iterable
+
+import numpy as np
+
+from cirq import protocols
+from cirq.ops import raw_types
+
+
+class AsymmetricDepolarizingChannel(raw_types.Gate):
+    """A channel that depolarizes asymmetrically along different directions."""
+
+    def __init__(self, p_x: float, p_y: float, p_z: float) -> None:
+        """The asymmmetric depolarizing channel.
+
+        This channel applies one of the three Pauli operators at random,
+        or does nothing.  The supplied probabilities must be valid probabilities
+        and the sum p_x + p_y + p_z must be a valid probability or else this
+        constructor will raise a ValueError.
+
+        This channel evolves a density matrix via
+            \rho -> (1 -p_x + p_y + p_z) \rho
+                    + p_x X \rho X + p_y Y \rho Y + p_z Z \rho Z
+
+        Args:
+            p_x: The probability that a Pauli X error occurs.
+            p_y: The probability that a Pauli Y error occurs.
+            p_z: The probability that a Pauli Z error occurs.
+
+        Raises:
+            ValueError if the args or the sum of the args are not probabilities.
+        """
+        def validate_probability(p, p_str):
+            if p < 0:
+                raise ValueError('{} was less than 0.'.format(p_str))
+            elif p > 1:
+                raise ValueError('{} was greater than 1.'.format(p_str))
+            return p
+        self._p_x = validate_probability(p_x, 'p_x')
+        self._p_y = validate_probability(p_y, 'p_y')
+        self._p_z = validate_probability(p_z, 'p_z')
+        self._p_i = 1 - validate_probability(p_x + p_y + p_z, 'p_x + p_y + p_z')
+
+    def _channel_(self) -> Iterable[np.ndarray]:
+        return (
+            np.sqrt(self._p_i) * np.eye(2),
+            np.sqrt(self._p_x) * np.array([[0, 1], [1, 0]]),
+            np.sqrt(self._p_y) * np.array([[0, -1j], [1j, 0]]),
+            np.sqrt(self._p_z) * np.array([[1, 0], [0, -1]])
+        )
+
+    def _eq_tuple(self):
+        return (AsymmetricDepolarizingChannel,
+                self._p_i, self._p_x, self._p_y, self._p_z)
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self._eq_tuple() == other._eq_tuple()
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __repr__(self) -> str:
+        return (
+            '(cirq.AsymmetricDepolarizingChannel(p_x={!r},p_y={!r},p_z={!r}))'
```

Comment:
The outside parens in the repr are unnecessary.

Reply:
OK.  There are a bunch of places where there are ()s in reprs.

Reply:
The general rule of thumb is "if it's of the form X**Y then you need them, because then appending `.stuff()` won't work". Use `cirq.testing.assert_equivalent_repr` to validate that a parens-less repr is correct.

Reply:
You can specify this repr and str in terms of `asymmetric_depolarize` if you want.

Reply:
Huh it looks like my change didn't get pushed.